lib/global_id_utils/locator.rb

```diff
@@ -21,12 +21,28 @@ def self.model_class_for(gid)
     private_class_method :model_class_for
 
     def self.find_records(model_class, ids, options)
+      ids = [ids] unless ids.is_a?(Array)
+      ids = ids.compact.uniq
+
       unscoped(model_class) do
-        if options[:ignore_missing]
-          model_class.where(model_class.gid_model_id_attribute => ids)
-        else
-          find(model_class, ids)
+        rel = model_class.where(model_class.gid_model_id_attribute => ids)
+
+        unless options[:ignore_missing]
+          found_ids = rel.pluck(model_class.gid_model_id_attribute)
+          not_found_ids = ids.map(&:to_s) - found_ids.map(&:to_s)
+
+          if ids.count != found_ids.count
+            rel.raise_record_not_found_exception!(
+              ids,
+              found_ids.count,
+              ids.count,
+              model_class.gid_model_id_attribute,
+              not_found_ids,
+            )
```

Comment:
Here is the source for #raise_record_not_found_exception!:

https://github.com/rails/rails/blob/b9ca94caea2ca6a6cc09abaffaad67b447134079/activerecord/lib/active_record/relation/finder_methods.rb#L327-L353